References - <br> 
> https://stackoverflow.com/questions/50747947/embedding-in-pytorch <br>
> https://stats.stackexchange.com/questions/270546/how-does-keras-embedding-layer-work
> https://stackoverflow.com/questions/49466894/how-to-correctly-give-inputs-to-embedding-lstm-and-linear-layers-in-pytorch

## Embedding Layer <br>  
* Embedding layer is a vector representation of all the words or sentences or characters. It act as a lookup table. By lookup table I mean assuming that each index/number represent a word so each number will have a vector in the lookup table. <br>
* So what is the argument of embedding layer, what is the input of embedding layer what it gives us as an output  ? <br>
> Input arguments - vocab size ( number of unique words if we are considering word embedding or unqiue sentence if we are considering sentence embedding), vector size ( or the embedding size which it will give us as an output) <br>
> Input - batch size or list ( where list act as the batch uske andar hamare sentences hai because of that sequence length hai next argument or dimension) , sequence length <br>
> Output - Batch size,sequence length, vector size <br> 

Lets understand it with an example

In [1]:
ex1 = 'This tutorial is amazing'
ex2 = 'I know it is'

list_of_unique_words = <b>[This,tutorial, is amazing, I know, it] </b><br>

vocab_size = length of this list = <b> length(list_of_unique_words)  </b><br>

vocab_size = 7 <br> 

vector_size = 3 (it is given by user) 

batch_size = length of total number of sentences we are passing, here we have 2 sentences. <br> 
batch_size = 2 <br> 

sequence_length = length of biggest sentence. In our example it is 4. <br>
sequence_length = 4 <br>

So from above things our output should have following dimension -  <br> 
Output - 2, 4,3 <br> 

Before starting we have to convert our each word into a number because computer can't understand strings. <br> 
So for simplicity lets assign them numbers in the following manner - <br> 

This - 0 <br>
tutorial - 1 <br>
is - 2 <br>
amazing - 3 <br> 
I - 4 <br> 
know - 5 <br> 
it - 6 <br> 

our new examples will look like this - 

ex1 = [0,1,2,3] <br> 
ex2 = [4,5,6,2] <br> 

The batch will be  <br>  
b = [ex1,ex2]

### In Summary we can say Embedding takes following input and returns following output   <br>
> ####  Params --> (vocab_size (unique list of words), embedding vector)
> ####  Input --> (batch_of_sentences or batch and max_sequence length from matrix prespective) 
> ####  Output --> (batch_size,max_sequence_length( maximum length of input), embedding vector)

In [2]:
import torch 
import torch.nn as nn 

In [9]:
b  = torch.LongTensor([[0,1,2,3],[4,5,6,2]])
embedding = nn.Embedding(7,3)
output = embedding(b)
output

tensor([[[-0.1544, -0.1875,  1.0524],
         [ 1.2851, -0.1237, -1.4782],
         [-1.7223,  2.0916, -1.4331],
         [ 0.2040, -0.5076, -0.1357]],

        [[ 0.2340,  0.0438, -1.1761],
         [ 0.3264, -0.1366,  0.7144],
         [-0.1838, -0.2113, -0.7216],
         [-1.7223,  2.0916, -1.4331]]], grad_fn=<EmbeddingBackward>)

In [6]:
output.shape

torch.Size([2, 4, 3])

Embedding vector lookup looks like this - <br>

This - 0 - -0.1544, -0.1875,  1.0524 <br> 
is - 1 - 1.2851, -0.1237, -1.4782 <br>  

So each index represent each value and there embedding too.

In [11]:
#Way to display image in markdown
#The pictures are from colah blog.
#![RNN Layer](https://colah.github.io/posts/2015-08-Understanding-LSTMs/img/RNN-rolled.png=1550)

## RNN Layer

### RNN cell looks like this 
<img src="https://colah.github.io/posts/2015-08-Understanding-LSTMs/img/RNN-rolled.png" width="100" height="100">

### RNN Layer Looks like this :
<img src="https://colah.github.io/posts/2015-08-Understanding-LSTMs/img/RNN-unrolled.png" width="500" height="200">

As we can see from above diagram RNN cell takes 2 inputs and returns 2 outputs.  <br>
> *  The input are input sequence and hidden vector. <br>
> *  The output is hidden vector and output cell

In [10]:
torch.randn(5, 3, 10)

tensor([[[-4.8988e-01,  6.1117e-01, -1.2516e+00, -2.6599e-01,  1.6506e+00,
           1.4007e+00,  1.5334e-01,  3.2586e-01, -9.3827e-01, -1.8486e+00],
         [-3.3065e-01,  5.2652e-01,  1.2147e+00, -1.1235e+00,  6.1428e-01,
          -2.7109e-01,  9.4458e-01, -6.1269e-06, -9.6571e-01,  5.9961e-01],
         [ 4.6273e-01,  3.8383e-01,  9.8650e-01,  1.4029e-01, -1.2388e+00,
           2.2979e+00,  3.9534e-01,  4.7278e-01,  1.1889e+00, -4.1915e-01]],

        [[-8.8662e-01,  1.0195e+00,  5.7871e-01,  1.0204e+00, -1.5576e-01,
           4.2497e-01,  7.4028e-01,  8.4036e-01,  3.2050e-01,  7.3251e-01],
         [-1.1737e-01, -1.2300e+00,  7.2090e-01, -2.7959e-01,  3.0336e-01,
          -1.1116e+00,  1.2790e-01,  1.1594e+00, -2.1750e-01, -2.2292e+00],
         [ 1.8969e+00,  1.8047e+00,  1.3417e-01, -1.3703e+00, -7.1033e-01,
           1.2285e+00, -9.6645e-01,  1.1403e+00, -1.0947e+00, -1.2524e+00]],

        [[ 5.2239e-01, -1.0371e-01,  1.3721e+00,  4.1747e-01, -1.7231e+00,
          -1.40

In [13]:
#batch = 2, seq= 2, input= 2
inputs = torch.randn(2,2,2)

In [14]:
#number_layers * direction = 1 layer * 1 direction , batch=  2 , vector = 3
hidden = torch.randn(1,2,3)

In [15]:
rnn = torch.nn.RNN(2,3,1,batch_first=True)

In [16]:
outputs,hiddens = rnn(inputs,hidden)

In [17]:
outputs.shape

torch.Size([2, 2, 3])

In [18]:
hiddens.shape

torch.Size([1, 2, 3])

In [23]:
outputs

tensor([[[-0.3622,  0.3926,  0.6566],
         [-0.3643,  0.4057,  0.2548]],

        [[-0.4595,  0.5373,  0.2882],
         [-0.3626,  0.3899,  0.1920]]], grad_fn=<TransposeBackward1>)

In [24]:
output.shape

torch.Size([2, 4, 3])

In [29]:
#batch,b
#outputs.view(seq_len, batch, num_directions, hidden_size)
outputs.view(2, 2, 1, 3)

tensor([[[[-0.3622,  0.3926,  0.6566]],

         [[-0.3643,  0.4057,  0.2548]]],


        [[[-0.4595,  0.5373,  0.2882]],

         [[-0.3626,  0.3899,  0.1920]]]], grad_fn=<ViewBackward>)

In [31]:
hiddens

tensor([[[-0.3643,  0.4057,  0.2548],
         [-0.3626,  0.3899,  0.1920]]], grad_fn=<StackBackward>)

In [32]:
hidden

tensor([[[ 0.4343, -1.1487, -0.6000],
         [ 1.4566, -0.7191,  1.2151]]])

In [33]:
inputs

tensor([[[-1.0572,  0.4944],
         [ 1.0124, -0.2122]],

        [[-0.0909, -0.2828],
         [ 0.1539,  0.0042]]])

In [34]:
hiddens.shape

torch.Size([1, 2, 3])

In [35]:
#h_n.view(num_layers, num_directions, batch, hidden_size).
hiddens.view(1,1,2,3)

tensor([[[[-0.3643,  0.4057,  0.2548],
          [-0.3626,  0.3899,  0.1920]]]], grad_fn=<ViewBackward>)

In [41]:
out = outputs.view(1,2,2,3)

In [42]:
out

tensor([[[[-0.3622,  0.3926,  0.6566],
          [-0.3643,  0.4057,  0.2548]],

         [[-0.4595,  0.5373,  0.2882],
          [-0.3626,  0.3899,  0.1920]]]], grad_fn=<ViewBackward>)

In [50]:
outputs.view(2,2,3)

tensor([[[-0.3622,  0.3926,  0.6566],
         [-0.3643,  0.4057,  0.2548]],

        [[-0.4595,  0.5373,  0.2882],
         [-0.3626,  0.3899,  0.1920]]], grad_fn=<ViewBackward>)

In [62]:
outputs[0][1]

tensor([-0.3643,  0.4057,  0.2548], grad_fn=<SelectBackward>)

In [57]:
hiddens

tensor([[[-0.3643,  0.4057,  0.2548],
         [-0.3626,  0.3899,  0.1920]]], grad_fn=<StackBackward>)

Hindi mein isko thoda lete hai. Kitne batch hai. Ek batch mein kitne words hai and un words ko agar vector space mein represent kre to kaise rehga.<br>

So when I said batch size of 2 matlab 2 batch lene hai and sequence length of 2 matlab ek batch mein max sequence itni words rehengs and input_size or vector size mane vector space mein usko itne vector value represent krenge. 

<i> So ye jo kr reh h vo bht simple hai ek misundertanding apan se ho gayi. So when I say batch matlab vo ek bar mein jayega. Once one batch has completed tab dusra batch jayega. So agar hum upar ke outputs dekhe to 2 batches hai. Phela batch aya. Ab maximum sequence length hai 2. So index 1 position ke value hidden value hiddens ne store kr liye and output ne sab kch store kr liya and ne </i>

This is how pytorch stores vector in hidden and output layer. 

<img src="https://miro.medium.com/max/640/1*tUxl5-C-t3Qumt0cyVhm2g.png" width="500" height="200">

## References on RNN 
> * https://towardsdatascience.com/pytorch-basics-how-to-train-your-neural-net-intro-to-rnn-cb6ebc594677 
> * https://towardsdatascience.com/recurrent-neural-networks-d4642c9bc7ce
> * https://datascience-enthusiast.com/DL/Building_a_Recurrent_Neural_Network-Step_by_Step_v1.html
